versioncontrol/peer_test.go

```diff
@@ -0,0 +1,217 @@
+
+		// test invalid rollout for each binary
+		for i := 1; i < fieldCount; i++ {
+			versions := versioncontrol.Versions{}
+			versionsValue := reflect.ValueOf(&versions)
+			field := reflect.Indirect(versionsValue).Field(i)
+
+			binary := versioncontrol.Binary{
+				Rollout: versioncontrol.Rollout{
+					Seed:   "",
+					Cursor: 0,
+				},
+			}
+
+			field.Set(reflect.ValueOf(binary))
+
+			config := versioncontrol.Config{
+				Versions: testServiceVersions,
+				Binary:   versions,
+			}
+
+			peer, err := versioncontrol.New(zaptest.NewLogger(t), &config)
+			require.NoError(t, err)
+			require.NotNil(t, peer)
+		}
+	})
+}
+
+func TestPeer_Run_error(t *testing.T) {
+	for _, scenario := range rolloutErrScenarios {
+		scenario := scenario
+		t.Run(scenario.name, func(t *testing.T) {
+			versionsType := reflect.TypeOf(versioncontrol.Versions{})
+			fieldCount := versionsType.NumField()
+
+			// test invalid rollout for each binary
+			for i := 1; i < fieldCount; i++ {
+				versions := versioncontrol.Versions{}
+				versionsValue := reflect.ValueOf(&versions)
+				field := reflect.Indirect(versionsValue).Field(i)
+
+				binary := versioncontrol.Binary{
+					Rollout: scenario.rollout,
+				}
+
+				field.Set(reflect.ValueOf(binary))
+
+				config := versioncontrol.Config{
+					Binary: versions,
+				}
+
+				peer, err := versioncontrol.New(zaptest.NewLogger(t), &config)
+				require.Nil(t, peer)
+				require.Error(t, err)
+				require.Contains(t, err.Error(), scenario.errContains)
+			}
+		})
+	}
+}
+
+func TestVersions_ValidateRollouts(t *testing.T) {
+	versions := validRandVersions(t)
+	err := versions.ValidateRollouts(zaptest.NewLogger(t))
+	require.NoError(t, err)
+}
+
+func TestRollout_Validate(t *testing.T) {
+	for i := 0; i < 100; i++ {
+		rollout := versioncontrol.Rollout{
+			Seed:   randSeedString(t),
+			Cursor: i,
+		}
+
+		err := rollout.Validate("test", zaptest.NewLogger(t))
+		require.NoError(t, err)
+	}
+}
+
+func TestRollout_Validate_error(t *testing.T) {
+	for _, scenario := range rolloutErrScenarios {
+		scenario := scenario
+		t.Run(scenario.name, func(t *testing.T) {
+			err := scenario.rollout.Validate("test", zaptest.NewLogger(t))
+			require.Error(t, err)
+			require.True(t, versioncontrol.RolloutErr.Has(err))
+			require.Contains(t, err.Error(), scenario.errContains)
+		})
+	}
+}
+
+func validRandVersions(t *testing.T) versioncontrol.Versions {
+	return versioncontrol.Versions{
+		Satellite: versioncontrol.Binary{
+			Rollout: randRollout(t),
+		},
+		Storagenode: versioncontrol.Binary{
+			Rollout: randRollout(t),
+		},
+		Uplink: versioncontrol.Binary{
+			Rollout: randRollout(t),
+		},
+		Gateway: versioncontrol.Binary{
+			Rollout: randRollout(t),
+		},
+		Identity: versioncontrol.Binary{
+			Rollout: randRollout(t),
+		},
+	}
+}
+
+func randRollout(t *testing.T) versioncontrol.Rollout {
+	return versioncontrol.Rollout{
+		Seed:   randSeedString(t),
+		Cursor: rand.Intn(100),
```

Comment:
`rand.Intn(100)` will generate random number between 0-99. It should be `rand.Intn(101)`, so we have the chance to test `cursor = 100`.